frame/gift/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+//! balance to Bob's final account.
+//!
+//! ## Design
+//!
+//! This pallet is designed to solve the following problems:
+//! * Allow Alice to send gift to a user who does not have an account (yet)
+//! * Allow a new user to safely claim those funds without exposing or involving anyone else for
+//!   account creation
+//! * Allow the new user to make that claim without needing to have any initial funds (a free
+//!   transaction)
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::{
+		pallet_prelude::*,
+		traits::{Currency, ExistenceRequirement, ReservableCurrency, OnTransfer},
+		sp_runtime::traits::{CheckedAdd, Saturating, Zero},
+	};
+	use frame_system::pallet_prelude::*;
+
+	type BalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// The Balances of your system.
+		type Currency: ReservableCurrency<Self::AccountId>;
+		/// The additional deposit needed to place a gift. Should be greater than the existential
+		/// deposit to avoid killing the gifter account.
+		type GiftDeposit: Get<BalanceOf<Self>>;
+		/// The minimum gift amount. Should be greater than the existential deposit.
+		type MinimumGift: Get<BalanceOf<Self>>;
+		/// Pallet hook for when a gift is claimed.
+		type OnGift: OnTransfer<Self::AccountId, BalanceOf<Self>>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[derive(Encode, Decode)]
+	pub struct GiftInfo<AccountId, Balance> {
+		gifter: AccountId,
+		deposit: Balance,
+		amount: Balance,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn gifts)]
+	pub type Gifts<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId, // Gift Controller
+		GiftInfo<T::AccountId, BalanceOf<T>>,
+	>;
+
+	#[pallet::event]
+	#[pallet::metadata(T::AccountId = "AccountId", BalanceOf<T> = "Balance")]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A gift has been created! [amount, controller]
+		GiftCreated(BalanceOf<T>, T::AccountId),
+		/// A gift has been claimed! [controller, amount, to]
+		GiftClaimed(T::AccountId, BalanceOf<T>, T::AccountId),
+		/// A gift has been removed :( [to]
+		GiftRemoved(T::AccountId),
+	}
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// User already has a pending gift.
+		PendingGift,
+		/// Don't be cheap... Get your friend a good gift!
+		GiftTooSmall,
+		/// An overflow has occurred.
+		Overflow,
```

Comment:
Now we have `ArithmeticError`, imo it should be preferred than custom overflow error. :] https://github.com/paritytech/substrate/blob/7fe74c4563a02a37b8546a378625e8fee1e8056b/primitives/runtime/src/lib.rs#L472